lib/src/main/java/com/auth0/android/lock/internal/configuration/Connection.java

```diff
@@ -0,0 +1,186 @@
+package com.auth0.android.lock.internal.configuration;
+
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static com.auth0.android.util.CheckHelper.checkArgument;
+
+public class Connection implements BaseConnection, DatabaseConnection, OAuthConnection, PasswordlessConnection {
+
+    private String strategy;
+    private String name;
+    private Map<String, Object> values;
+    private int minUsernameLength;
+    private int maxUsernameLength;
+
+    private Connection(@NonNull String strategy, Map<String, Object> values) {
+        checkArgument(values != null && values.size() > 0, "Must have at least one value");
+        final String name = (String) values.remove("name");
+        checkArgument(name != null, "Must have a non-null name");
+        this.strategy = strategy;
+        this.name = name;
+        this.values = values;
+        parseUsernameLength();
+    }
+
+    @Override
+    public String getName() {
+        return name;
+    }
+
+    @Override
+    public String getStrategy() {
+        return strategy;
+    }
+
+    /**
+     * Getter for the Type of Connection
+     *
+     * @return the connection Type.
+     */
+    int getType() {
+        switch (strategy) {
+            case "auth0":
+                return AuthType.DATABASE;
+            case "sms":
+            case "email":
+                return AuthType.PASSWORDLESS;
+            case "ad":
+            case "adfs":
+            case "auth0-adldap":
+            case "custom":
+            case "google-apps":
+            case "google-openid":
+            case "ip":
+            case "mscrm":
+            case "office365":
+            case "pingfederate":
+            case "samlp":
+            case "sharepoint":
+            case "waad":
+                return AuthType.ENTERPRISE;
+            default:
+                return AuthType.SOCIAL;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> T getValueForKey(String key) {
+        return (T) this.values.get(key);
```

Comment:
This might cause `ClassCastException` we need to check if it can be casted first